src/net.cpp

```diff
@@ -1681,8 +1684,10 @@ void CConnman::ThreadOpenConnections()
             if (nANow - addr.nLastTry < 600 && nTries < 30)
                 continue;
 
-            // only consider nodes missing relevant services after 40 failed attempts and only if less than half the outbound are up.
-            if ((addr.nServices & nRelevantServices) != nRelevantServices && (nTries < 40 || nOutbound >= (nMaxOutbound >> 1)))
+            // only consider nodes missing relevant services we have at least 4 already,
+            // and at least 6 once SegWit has activated
+            if ((addr.nServices & nRelevantServices) != nRelevantServices && (nTries < 40 &&
+                    (nRelevant < 5 || (fWitnessActive && nRelevant < 7))) && !fFeeler)
```

Comment:
Several problems with this:
* You're reading fWitnessActive here without holding cs_main, so you have a potential data race (even though in this particular case, likely every supported platform will actually do what you'd expect, this is undefined behaviour according to the C++ standard, and the compiler is allowed to make this do anything - including deleting your files). To correctly write what you're trying to do here, you'd need to make fWitnessActive an `std::atomic<bool>` instead.
* Having an fWitnessActive in the network code is a huge layer violation. The network code shouldn't need to know about such concepts.
* As I've explained, we need the connectivity with witness nodes at the time the activation occurs - which means the preferential peering is most important *before* the actual activation. The easiest way to accomplish that is by doing it always.